objectives.md

```diff
@@ -0,0 +1,75 @@
+# NGINX Ingress Controller Service Level Objectives and Shared Responsibilities with Customers
+
+Last updated: 21. Oct 2020
+
+To supplement our [general Service Level Objectives](https://www.giantswarm.io/blog/the-radical-way-giant-swarm-handles-service-level-objectives), Giant Swarm and its customers have a shared objective to keep the NGINX Ingress Controller app up and running.
+
+This doc clarifies what Giant Swarm and customer need to do in order to accomplish that.
+
+In essence:
+
+*   **Giant Swarm** (1) monitors and alerts on basic metrics, (2) provides release quality assurance, (3) makes upgrades available w/in 30 days of upstream, and (4) communicates customer-impacting changes, provides migration paths, etc. 
+*   **Customer** takes care of (1) fine-grained and workload-specific monitoring, and capacity and performance management, (2) testing upgrades when jumping over multiple releases, (3) reading release notes carefully with Solution Engineers.
+
+## Giant Swarm’s responsibility
+
+**First, monitoring and alerting.**
+
+In production, we monitor and alert on these basic metrics:
+*   For each workload, when at least one pod was unavailable for more than 5 minutes. In addition, when the error budget’s burn rate
+    *   Was in the last 10mins, higher than 1%
+    *   Was in the last hour, higher than the max allowable rate
+    *   Will run out in ~3 days
+*   When a container crashes repeatedly.
+*   When an app fails to install or upgrade. We get notified, but not paged yet.
+
+**Second, Quality Assurance of releases to reduce likelihood of introducing regressions.**
+
+Performance Tests
+*   Setup: We start a 3-node cluster of m5.xlarge Instance Type on AWS. We use affinity to start three deployments: (1) NGINX IC app, (2) StormForger Load Testing app, and (3) Gatling Performance Testing tool.
+*   Generated load
+    *   Gatling is configured to hit Stormforger through NGINX IC 
+    *   Generated load starts with 150 concurrent users. Over 15 seconds, load ramps up to 500 concurrent users. Each concurrent user issues 100 echo requests to send 50.000 requests in total.
+*   Assertion
+    *   Request success ratio (HTTP 200 response code) is >= 99.5% of the requests
+    *   Throughput is more than 2000 requests per second
+
+General Quality Assurance
+*   Static Code Analysis (linting, etc.)
+*   Integration Tests to ensure that the Helm chart can be deployed.
+*   Functional Tests
+    *   Platform Compatibility Testing, with latest platform releases across providers (AWS, Azure, and KVM)
+    *   End-to-End smoke tests to make sure Ingress traffic gets routed when (1) installing a new app, and (2) upgrading an app
```

Comment:
Interested readers will want to know more about these topics. Can we at least link to the CI configuration or something similar?

Reply:
yeah, that will be good. we discussed wanting to do that before as well :) 

Is it visible to non-GS people? Can you suggest sth? ping @piontec

Reply:
[This file](https://github.com/giantswarm/nginx-ingress-controller-app/blob/master/test/kat/test_nginx_ic.py) contains the testing logic. (Spin up the additional apps, test against them)

[This file](https://github.com/giantswarm/nginx-ingress-controller-app/blob/master/.circleci/config.yml) contains test orchestration (run on AWS)

Reply:
thanks!! @ubergesundheit , do you mind editing this PR to include these links? I'm not sure what they are and how to include them. 

After that, you can approve and I can merge already haha

Reply:
```suggestion
Performance Tests
*   Setup: We start a 3-node cluster of m5.xlarge Instance Type on AWS. We use affinity to start three deployments: (1) NGINX IC app, (2) StormForger Load Testing app, and (3) Gatling Performance Testing tool.
*   Generated load
    *   Gatling is configured to hit Stormforger through NGINX IC 
    *   Generated load starts with 150 concurrent users. Over 15 seconds, load ramps up to 500 concurrent users. Each concurrent user issues 100 echo requests to send 50.000 requests in total.
*   Assertion
    *   Request success ratio (HTTP 200 response code) is >= 99.5% of the requests
    *   Throughput is more than 2000 requests per second

General Quality Assurance
*   Static Code Analysis (linting, etc.)
*   Integration Tests to ensure that the Helm chart can be deployed.
*   Functional Tests
    *   Platform Compatibility Testing, with latest platform releases across providers (AWS, Azure, and KVM)
    *   End-to-End smoke tests to make sure Ingress traffic gets routed when (1) installing a new app, and (2) upgrading an app

Implementation details regarding performance tests and general quality assurance can be found in [performance test files](https://github.com/giantswarm/nginx-ingress-controller-app/blob/master/test/kat), [integration test files](https://github.com/giantswarm/nginx-ingress-controller-app/tree/master/integration) and [CircleCI configuration](https://github.com/giantswarm/nginx-ingress-controller-app/blob/master/.circleci/config.yml).

```